### PR TITLE
CBL-2694: Prevent rev list gaps

### PR DESCRIPTION
**Pull request by callumbirks**
With this change, we will now keep ~100~ 50 revs in the history of a doc, rather than 20. This shouldn't make a huge impact on storage requirements, because we only keep bodies for the current rev, and the last rev we synced with SG.

~With this current implementation, if the client has somehow made > 100 revisions since they last synced, SG will receive a rev list with gap. I can change this so we give SG the random revIDs we generate (I've tested the random revIDs, and they're fine). But I'm not sure if we want to do that, or just allow the warning to happen for > 100 revisions because if a customer is making > 100 revisions between syncs frequently, that may be a sign of another problem.~

With this implementation, if the client has made > 50 revisions since they last synced, the gap in the history to be sent to SG will be filled with randomly generated revIDs from `c4doc_getRevisionHistory()`. The client will also be warned of the large amount of mutations between sync, if we had to use random revIDs. We will send SG a history of up to 100 revs (50 real, 50 randomly generated). If there is more than 100 revisions since last sync, SG will receive a rev list with a gap, and therefore SG will warn the user about a gap.

### CODE REVIEW

**Comment by borrrden** [comment]
It doesn't have to be either or, we should do both the random IDs AND a warning that we are doing something funky.

File: Replicator/RevFinder.cc [hunk 1]
```diff
@@ -122,7 +122,7 @@ namespace litecore::repl {
                 response.compressed = true;
                 if (!_db->usingVersionVectors()) {
 #if 1
-                    response["maxHistory"_sl] = 20;
+                    response["maxHistory"_sl] = 100;
```
Comment (borrrden) [inline Replicator/RevFinder.cc:125]: Is it possible to get this value from ReplicatorTuning.hh ?

Comment (callumbirks) [reply Replicator/RevFinder.cc:125]: You're right, it is.

**Comment by callumbirks** [comment]
I've changed the revTree depth to 50, and left the replication maxHistory at 100. These numbers can be discussed if we think something else is suitable.
I've added a log warning for a gap in the revision history, with some (hopefully) helpful info for the client.

**Comment by callumbirks** [comment]
With these numbers (revTree depth 50, max history 100): above 50 revisions since last sync, we will send SG random revIDs and there will be a warning logged to the client. Above 100 revisions, there will also be a gap in the rev list sent to SG, so a warning will appear there as well.

**Review by pasin — COMMENTED** [review]
I'm not sure about changing the max depth = 20 if it doesn't cause the issue.

**Comment by callumbirks** [comment]
> I'm not sure about changing the max depth = 20 if it doesn't cause any issues.

So the original issue is SG receiving a rev list with gaps. Increasing the `maxHistory` for the Replicator means we will generate random revIDs to fill in the gap in the rev history. This stops SG from receiving a rev list with gaps (up to the `maxHistory`). However, the concern is that because the gap is filled with randomly generated revIDs, this causes conflicts with other clients' rev history.

The revTree depth is not as storage-expensive as it was in the past, because now we only store bodies of the current rev, and the last rev we synced with remote. And increasing the revTree depth will reduce the number of cases where we have conflicts with another client.

I have written a full test case to test the conflicts, and they seem to resolve fine, but the less the better.

**Review by snej — APPROVED** [review]
Not fully comfortable with storing more revIDs locally, but the performance impact probably is negligible...

File: Replicator/tests/ReplicatorCollectionSGTest.cc [hunk 2]
```diff
@@ -1889,3 +1889,133 @@ TEST_CASE_METHOD(ReplicatorCollectionSGTest, "Pull iTunes deltas from Collection
     C4Log("-------- %.3f sec with deltas, %.3f sec without; %.2fx speed",
           timeWithDelta, timeWithoutDelta, timeWithoutDelta/timeWithDelta);
 }
+
+TEST_CASE_METHOD(ReplicatorCollectionSGTest, "Give SGW random rev history and conflicts", "[.CBL-2694]") {
+    // The idea of this test is to exceed the RevTree history limit, so we can see what happens when we use the
```
Comment (jianminzhao) [inline Replicator/tests/ReplicatorCollectionSGTest.cc:1894]: I am not sure what's the motivation of the changes to the two constants, 20 -> 50 and 20 -> 100. I restored that changes, and this test also passes.

**Review by jianminzhao — CHANGES_REQUESTED** [review]

File: LiteCore/Database/DatabaseImpl.cc [hunk 3]
```diff
@@ -51,7 +51,7 @@ namespace litecore {
 
 
     static constexpr slice kMaxRevTreeDepthKey = "maxRevTreeDepth";
-    static constexpr uint32_t kDefaultMaxRevTreeDepth = 20;
+    static constexpr uint32_t kDefaultMaxRevTreeDepth = 100;
```
[the remark this answers is not in the document]
Comment (callumbirks) [reply LiteCore/Database/DatabaseImpl.cc]: The max revTree depth is how many revs to store in the history of the revTree document, the Replicator's maxHistory is the max depth of history to send to SG during sync. We have a facility in `c4doc_getRevisionHistory` to generate random revIDs to fill in the gap between `maxRevTreeDepth` and Replicator's `maxHistory`, which seems like it was intended to be used based on the comments in the function, but never was because previously the values of `maxRevTreeDepth` and `maxHistory` were the same.

File: Replicator/RevFinder.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (callumbirks) [reply Replicator/RevFinder.cc]: It does in the most recent commit

File: Replicator/tests/ReplicatorCollectionSGTest.cc (hunk 2 above)
Comment (callumbirks) [reply Replicator/tests/ReplicatorCollectionSGTest.cc:1894]: The test is intended to see the behaviour of conflict resolution when we are sending SG a history that contains randomly generated rev history. So the test just uses whatever the values are of `maxHistory` and `maxRevTreeDepth`, creates some conflicts in the rev history, and ensures that conflict resolution behaves as expected. 

Looking at the test now though, I do need to update it a bit since I changed the values from (100/20) to (100/50) since I first made the test.

**Comment by callumbirks** [comment]
I've updated the test and changed it to use `kDefaultMaxHistory` rather than a hardcoded number. I did a lot of refactoring to try and make the test neater.

File: LiteCore/Database/DatabaseImpl.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (callumbirks) [reply LiteCore/Database/DatabaseImpl.cc]: In the latest commit, `maxRevTreeDepth`/`maxHistory` are different (`50`/`100`)

File: Replicator/tests/ReplicatorCollectionSGTest.cc (hunk 2 above)
Comment (jianminzhao) [reply Replicator/tests/ReplicatorCollectionSGTest.cc:1894]: My wonder is actually the following. There are three parts of the in the PR.
1. Substance, it changed two constant, kDefaultMaxRevTreeDepth, kDefaultMaxHistory
2. Log the gap when the random revIDs are substituted in the history.
3. This test.
I tried to remove (1) and the test also passed. So, is the main object of PR is to add a test? Is there a thing that did not work without this PR?

Comment (callumbirks) [reply Replicator/tests/ReplicatorCollectionSGTest.cc:1894]: The test doesn't test something that's internally broken, the issue stems from SG receiving a rev history from us that has a gap. There were complaints from certain customers about this warning in SG. The test is simply to test that with the increased values, and thereby the usage of randomly generated revIDs, the conflict resolution will still behave as it does normally.

I discussed it with SG team, Jim and Jens and SG team considered whether they could just drop the warning from SG when they receive a gap. This is what Ben said to me:

> Adam replied saying this and then I pointed him to the litecore code generating non-deterministic revision IDs:
> > We wouldn't create fake conflicts if we had a deterministic approach for generating the intermediate digests (either clone the digest of the latest revision, or hash from parent rev ID).  Have we already reviewed whether the rev tree pruning algorithm (even for non-branching rev trees) is happy with gaps in generation numbers?

> It seems odd that the comment around that area of litecore code suggests its generating them in order to send them, but then never does

> It still feels like litecore should be filling that gap for us - couchdb (and therefore SG) formats revisions in a way that doesn't allow you to have gaps

So I decided the best course of action was to increase the length of history that we send to SG, so it is much harder to perform enough local mutations between syncs to end up sending a rev history with a gap to SG. That is why I propose to increase `ReplicatorTuning::kDefaultMaxHistory`.

The drawback with that, is any history in-between the revTree depth and maxHistory is randomly generated (within the function `TreeDocument::_getRevisionHistory`). As it is randomly generated, this could easily cause conflicts with other clients, and we want to cause as little conflicts as we can.

The current max revTree depth of 20 was set when we used to store the bodies of all revs in the tree, so each rev we store in the history was quite storage-expensive. As we now only store the bodies of the current rev and the last rev we synced with remote, additional revs in the revTree aren't too expensive (I'm assuming just however many bytes the revIDs are). 

Therefore, to reduce the usage of randomly generated revIDs, I want to increase `kDefaultMaxRevTreeDepth`, as it shouldn't be too storage-expensive and will significantly reduce the possibility of conflicts between clients' rev history.

Comment (callumbirks) [reply Replicator/tests/ReplicatorCollectionSGTest.cc:1894]: Additionally, the reason for logging a warning to the client when we had to use randomly generated revIDs is because if the client is making more than 50 mutations between syncs with SG, there is probably something going wrong.

Comment (jianminzhao) [reply Replicator/tests/ReplicatorCollectionSGTest.cc:1894]: So, the fundamental problem is that the client voluntarily limit the max-history. The protocol allows the reply to include "maxHistory." If it is absent, the client would voluntarily limit it to a number, which I assume is not part of the protocol. Now, we are in a bad situation. On one hand, we limit the history to, say, 20 or 100, but the last revision that the remote knows is deeper than the max. On the other hand, we want to include the last synced revision in the history. 
Either we don't self-impose the maxHistory, or the protocol allows gaps and SG should handle it. Of course, if SG puts maxHistory in the reply, it knows the history may be truncated.

Comment (jianminzhao) [reply Replicator/tests/ReplicatorCollectionSGTest.cc:1894]: `                if (hasRemoteAncestor(revID)) {
                    // Always write the common ancestor, making room if necessary:
                    if (revsWritten == maxRevs)
                        removeLast();
                    append(revID);
                    break;
                } else {`
^^^ from _getRevisionHistory

File: Replicator/tests/ReplicatorCollectionSGTest.cc [hunk 4]
```diff
@@ -1730,3 +1730,111 @@ TEST_CASE_METHOD(ReplicatorCollectionSGTest, "Pull iTunes deltas from Collection
     C4Log("-------- %.3f sec with deltas, %.3f sec without; %.2fx speed",
           timeWithDelta, timeWithoutDelta, timeWithoutDelta/timeWithDelta);
 }
+
+// This test may be used in future, if kDefaultMaxHistory > kDefaultMaxRevTreeDepth, and we wish to test
+// the conflict resolution behaviour for randomly generated rev history.
+#if 0
```
Comment (jianminzhao) [inline Replicator/tests/ReplicatorCollectionSGTest.cc:1736]: Why disable this test? It is still valuable to test sync in the presence of gaps, IMHO.

**Review by borrrden — APPROVED** [review]
I am also curious what happens when gaps are present, so tend to agree with Jianmin.

File: Replicator/tests/ReplicatorCollectionSGTest.cc [hunk 5]
```diff
@@ -1730,3 +1730,94 @@ TEST_CASE_METHOD(ReplicatorCollectionSGTest, "Pull iTunes deltas from Collection
     C4Log("-------- %.3f sec with deltas, %.3f sec without; %.2fx speed",
           timeWithDelta, timeWithoutDelta, timeWithoutDelta/timeWithDelta);
 }
+
+// Test replication and conflict resolution behaviour when we send SG a rev history with a gap (number of new
+// revs between syncs > repl::tuning::kDefaultMaxHistory).
+// This test attempts to emulate two separate clients with the same doc, where one client's rev history lands
+// in the gap of the other client's rev history.
+TEST_CASE_METHOD(ReplicatorCollectionSGTest, "Give SG a rev history with a gap", "[.SyncServerCollection]") {
```
Comment (jianminzhao) [inline Replicator/tests/ReplicatorCollectionSGTest.cc:1738]: I think. it's more interesting to demonstrate the difference between the current set of constants, 50/50, v.s. ideal ones, infinity/infinity, or 200/200 in your example. I know we cannot do it without modifying the source code, but perhaps you can shrinking the size of numRev1/numRev2 proportionally to fit to maxTreeDepth.

Comment (callumbirks) [reply Replicator/tests/ReplicatorCollectionSGTest.cc:1738]: I don't quite understand your meaning. Would you mean potentially having two Catch Sections for the test, one with numRev values that won't cause a gap, one with numRev values that will cause a gap? Then we can show that the behaviour is the same with or without a gap?

Comment (jianminzhao) [reply Replicator/tests/ReplicatorCollectionSGTest.cc:1738]: exactly

**Review by jianminzhao — APPROVED** [review]
